cmake/CMakeLists.txt

```diff
@@ -136,15 +136,15 @@ if((PKG_KOKKOS) AND (Kokkos_ENABLE_CUDA) AND NOT (CMAKE_CXX_COMPILER_ID STREQUAL
   set(CMAKE_TUNE_DEFAULT "${CMAKE_TUNE_DEFAULT} -Xcudafe --diag_suppress=unrecognized_pragma")
 endif()
 
-# we require C++11 without extensions. Kokkos requires at least C++14 (currently)
+# we require C++11 without extensions. Kokkos requires at least C++17 (currently)
 if(NOT CMAKE_CXX_STANDARD)
   set(CMAKE_CXX_STANDARD 11)
 endif()
 if(CMAKE_CXX_STANDARD LESS 11)
   message(FATAL_ERROR "C++ standard must be set to at least 11")
 endif()
-if(PKG_KOKKOS AND (CMAKE_CXX_STANDARD LESS 14))
-  set(CMAKE_CXX_STANDARD 14)
+if(PKG_KOKKOS AND (CMAKE_CXX_STANDARD LESS 17))
+  set(CMAKE_CXX_STANDARD 17)
```

Comment:
I think it would be better to do this on per-target basis with `set_target_properties()`, just in case somebody would like to compile lammps with C++20 or newer.

Reply:
We've had something like that before, but that led to all kinds of unexpected inconsistencies.
If you want C++-20 you have to use it for "everything" by setting `-DCMAKE_CXX_STANDARD=20`.

Reply:
see https://stackoverflow.com/questions/70566491/how-to-bump-the-c-standard-from-the-cmake-command-line

also Kokkos developer @dalg24 had some strong opinions on that.

Reply:
> If you want C++-20 you have to use it for "everything" by setting `-DCMAKE_CXX_STANDARD=20`.

On board with that approach.

Only unfortunate thing I see with the snippet above is if one explicitly configure with `-DCMAKE_CXX_STANDARD=11` or `14` it would silently be changed to `17` without even raising a warning.
You could do something like
```CMake
if(NOT CMAKE_CXX_STANDARD)
  if(PKG_KOKKOS)
    set(CMAKE_CXX_STANDARD 17)
  else()
    set(CMAKE_CXX_STANDARD 11)
  endif()
endif()

if(PKG_KOKKOS)
  if(CMAKE_CXX_STANDARD LESS 17)
    message(FATAL_ERROR "C++ standard must be set to at least 17")
  endif()
else()
  if(CMAKE_CXX_STANDARD LESS 11)
    message(FATAL_ERROR "C++ standard must be set to at least 11")
  endif()
endif()
```

Reply:
That stackoverflow issue is setting the C++ standard unconditionally. The quoted cmake script code from LAMMPS does not. It will *raise* the C++ standard to 14 (old) or 17 (new) is it has not been set or set to a lower value.

Reply:
@dalg24 True, but for now I am very happy with a solution that requires to write and maintain only 2 lines and wait until somebody runs into a situation where it creates a problem before considering a more complex approach.